### PR TITLE
NPC clearance heatmaps + central viz library & ego-frame kinematics fixes

### PR DESCRIPTION
**Pull request by danielsanchezaran**
## Summary

Two related strands of work on the scenario-generation / visualization stack.

### 1. NPC clearance heatmaps
- Per-step clearance log (ego world pose + realized road-border / stopped-NPC / moving-NPC distances) written by the closed-loop sim.
- `tools/heatmap_npc_clearance.py`: separate stopped vs moving clearance heatmaps along the route (worst clearance per arc bin) + per-type bad-area folders.
- Parameterized dump neighbor-slot count (`dump_neighbor_count`, default 320).

### 2. Central viz library + ego-frame kinematics correctness
- **Central renderer:** extracted `render_scene_at_step` (+ draw helpers) out of the Scene Branch Editor into a new gradio-free `scenario_generation/scene_render.py`; the GUI and a new `viz_saved_scenes.py` import it instead of duplicating.
- **Footprint convention unified everywhere** to match `loss.py`/`reward.py`: ego box = rear-axle offset `(length-wheelbase)/2`; neighbors = centroid (`length/2`). Fixed ~10 tools that hand-rolled boxes with the wrong offset (viz_p4_recovery, viz_prism_compare, viz_kinematic_floored, viz_t1_crossing_debug, visualize_scenes, recovery_sim/ghost family, replay/simulate trajectory footprints, placed obstacles).
- `--ego_wheelbase` threaded through the ghost CLIs so ego footprints use the true wheelbase instead of a `length*0.65` heuristic.
- **`advance_scene_mpc`**: added the neighbor centroid↔rear-axle conversion that `advance_scene` already does (model-driven NPCs were biased ~wheelbase/2 forward).
- **`ego_current_state` dump**: store full velocity & acceleration magnitude on ego-x with `vy=ay=0` (base_link convention, matching real data); centripetal accel computed for tracking but never written to `ay`.
- Scene Branch Editor fixes: auto-convert 3-col neighbor futures to 4-col on save; unconditional json import in Simulate (fixed UnboundLocalError on sidecar-less NPZ sets); neighbor width slider 3→6m.
- `sampled_trajectories` init standardized to zeros for deterministic inference (match production C++ node).

## Test
- All touched modules compile + smoke-import.
- Footprint convention verified against `reward._build_ego_bbox_corners`.
- Round-trip + real-vs-sim dataset checks for `ego_current_state` (vy/ay = 0 in real data).
- Saved-scene viz regenerated and visually confirmed.

### CODE REVIEW

File: rlvr/closed_loop/batched_rollout.py [hunk 1]
```diff
@@ -18,6 +18,20 @@
...
+
+    Matches the C++ production node: randn * temperature (zeros when
+    temperature/noise_scale is 0).  Shape: (B, P, future_len + 1, 4).
+    """
+    if noise_scale > 0.0:
+        return noise_scale * torch.randn(B, P, future_len + 1, 4, device=device)
+    return torch.zeros(B, P, future_len + 1, 4, device=device)
 from torch import nn
 
 from exploration_policy.model import ExplorationPolicy
```
[the remark this answers is not in the document]
Comment (danielsanchezaran) [reply rlvr/closed_loop/batched_rollout.py]: Fixed in b8c0e3a — moved `make_initial_latent` below the import block so imports stay grouped (ruff isort passes).

File: scenario_generation/tools/viz_saved_scenes.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                 fontsize=11, color="#1f9e3a" if ok else "#cc2222", y=0.99)
+    return fig, (ok, summary, flags)
+
+
+def main():
+    p = argparse.ArgumentParser()
+    g = p.add_mutually_exclusive_group(required=True)
+    g.add_argument("--scene_dir")
+    g.add_argument("--scenes", help="JSON list of NPZ paths")
+    p.add_argument("--reward_config", required=True)
```
[the remark this answers is not in the document]
Comment (danielsanchezaran) [reply scenario_generation/tools/viz_saved_scenes.py:141]: Fixed in b8c0e3a — `--scenes` help now reads 'Path to a JSON file containing a list of NPZ paths'.

File: scenario_generation/replay.py [hunk 3]
```diff
@@ -240,6 +240,10 @@ class SpawnConfig:
     # When set, per-step observations are dumped as training-style NPZs to
     # this directory. GT future is zeroed out (ranked-SFT generates its own).
     dump_npz_dir: str | None = None
+    # Neighbor-slot count for the dumped NPZs. Defaults to the model's
+    # neighbor dimension (320). Set lower only when dumping for a model with
+    # a different neighbor dimension. Past and future are built at this count.
+    dump_neighbor_count: int = 320
```
[the remark this answers is not in the document]
Comment (danielsanchezaran) [reply scenario_generation/replay.py:246]: Fixed in 7e58173 — SpawnConfig.validate() now requires dump_neighbor_count >= 1 when dump_npz_dir is set (fails fast).

File: rlvr/autoresearch/tools/recovery_sim.py [hunk 4]
```diff
@@ -31,15 +31,15 @@
 import subprocess
 from pathlib import Path
 
+import matplotlib.transforms as mtransforms
```
[the remark this answers is not in the document]
Comment (danielsanchezaran) [reply rlvr/autoresearch/tools/recovery_sim.py]: Fixed in 7e58173 — removed the unused `matplotlib.transforms` import.

File: rlvr/autoresearch/tools/recovery_sim.py [hunk 5]
```diff
@@ -31,15 +31,15 @@
...
 import numpy as np
 import torch
+from diffusion_planner.model.diffusion_planner import Diffusion_Planner
+from diffusion_planner.utils.config import Config
 from matplotlib.collections import LineCollection
 from matplotlib.figure import Figure
 from matplotlib.patches import Rectangle
```
[the remark this answers is not in the document]
Comment (danielsanchezaran) [reply rlvr/autoresearch/tools/recovery_sim.py]: Fixed in 7e58173 — removed the unused `Rectangle` import.
